### PR TITLE
Update for the new GeoInterface

### PR DESCRIPTION
**Pull request by rafaqz**
See https://github.com/JuliaGeo/GeoInterface.jl/pull/33

There are some issues with how to fit the odd format of Shapefiles - mixed multipoligons with rings and holes - into the OGC standard and GeoInterface. We need to work out how to efficiently index holes with `gethole` when we have to manually check.

### CODE REVIEW

**Comment by visr** [comment]
Nice work! How do you see this in relation to #39?

**Comment by rafaqz** [comment]
Yeah. If we do that, we won't need this PR, because we will have the interface in GB anyway. So this is kind of temporary really.

But there are three unmerged GB PRs here now and I dont know where they are at or which to move forward. 

Im also not totally sure if the upfront conversion to GB types is the best approach. With shapefiles we know the bounding boxes of all objects and can use it that information to reduce work. But with GeometryBasics conversion all shapes will be sorted into rings/holes on load. That can be a huge and unnessary overhead if we only need to analyse a small region of a large shape file. On the other hand I can also see how annoying shapefiles are to fit into the GeoInterface.

**Comment by rafaqz** [comment]
Seems like this will kill the other GeometryBasics PRs here? https://github.com/JuliaGeometry/GeometryBasics.jl/pull/173

We need metadata for the bounding box, which would be a shame to lose.

**Comment by rafaqz** [comment]
This is working, just waiting on the new GeoInterfaceRecipes subpackage to be merged and registered for plot recipes here.

- files are broken into object types with relevent methods contained
- because polygons here are really multipoligons, there are some lazy objects that handle getting the polygons and rings from them.
- everything 3d is actually 3d now.
- we have `getring` in GeoInterface to skip the whole searching for exteriors and holes step, which is much faster for rasterizing and plotting.

@visr my long term plan is to replace this package with a 100% lazy version we spoke about. And call it Shapefiles.jl so we can just use `Shapefile(filename)` as an object. And make that object automatically act as a feature table.

But for now this is working, and should actually be much faster to load, plot and rasterize than before.

**Comment by visr** [comment]
Cool! Besides the new GeoInterface, what (if any) is breaking about this? Does e.g. the README quick start example still work without changes?

**Comment by rafaqz** [comment]
@evetion and @visr, do you have any ideas on what to do about 3d polygons and the geointerface. I guess we need a 3d `inpolygon` algorithm to separate exterior and interior rings?

**Comment by evetion** [comment]
You can still pretend that they're 2d, as they're not truly 3d, but more like 2.5d. So the inpolygon just works with the first two coordinates.

**Comment by rafaqz** [comment]
Ah ok thats what I've done anyway. So its the same as before, but the actual coordinates returned are 3d.

**Comment by visr** [comment]
Feel free to drop julia pre-1.6 here by the way.

File: README.md [hunk 1]
```diff
@@ -20,7 +20,7 @@ table = Shapefile.Table(path)
 geoms = Shapefile.shapes(table)
 
 # whole columns can be retrieved by their name
-table.Descriptio  # => Union{String, Missing}["Square with triangle missing", "Smaller triangle", missing]
+table.Description  # => Union{String, Missing}["Square with triangle missing", "Smaller triangle", missing]
```
Comment (visr) [inline README.md:23]: It really is just `Descriptio`, just open the dbf in a plain text editor.
http://switchfromshapefile.org/#10characters
:(

Comment (rafaqz) [reply README.md:23]: The horror

**Comment by rafaqz** [comment]
If added `convert` methods, which should also allow writing to shapefile from other formats with a reboot of #34

File: src/multipatch.jl [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+GI.geomtrait(::MultiPatch) = GI.GeometryCollectionTrait()
+GI.ngeom(geom::MultiPatch) = length(geom.parts)
+GeoInterface.ncoord(::MultiPatch) = 3
+
+# TODO implement the rest of this
+function GI.getgeom(::GI.GeometryCollectionTrait, geom::MultiPatch, i::Integer)
+    error("`getgeom` not implemented for `MultiPatch`: open a github issue at JuliaGeo/Shapefile.jl if you need this.")
+    # T = SHAPETYPE[geom.parttypes[i]]
+end
+function GI.subtrait(geom::MultiPatch)
```
Comment (evetion) [inline src/multipatch.jl:33]: This is meant to be an internal method in GI no?

File: src/multipatch.jl [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+end
+
+GI.geomtrait(::MultiPatch) = GI.GeometryCollectionTrait()
+GI.ngeom(geom::MultiPatch) = length(geom.parts)
+GeoInterface.ncoord(::MultiPatch) = 3
+
+# TODO implement the rest of this
+function GI.getgeom(::GI.GeometryCollectionTrait, geom::MultiPatch, i::Integer)
+    error("`getgeom` not implemented for `MultiPatch`: open a github issue at JuliaGeo/Shapefile.jl if you need this.")
+    # T = SHAPETYPE[geom.parttypes[i]]
```
Comment (evetion) [inline src/multipatch.jl:31]: Remove or explain commented lines.

File: src/table.jl [hunk 4]
```diff
@@ -145,4 +145,5 @@ Get a vector of the geometries in a shapefile `Table`, without any metadata.
 """
 shapes(t::Table) = shapes(getshp(t))
 
+GeoInterface.extent(t::Table) = GeoInterface.extent(getshp(t))
```
Comment (evetion) [inline src/table.jl:148]: Here we could also implement `isfeature`, `geometry` and `properties` right?

**Comment by evetion** [comment]
Because of the major restructuring, I can't really tell what's new code and old. I have a few small comments, otherwise LGTM.

File: src/multipatch.jl (hunk 2 above)
Comment (rafaqz) [reply src/multipatch.jl:33]: Yeah, I'll delete. The problem was that MultiPatch has a small list of unusual subtraits that I gave up on trying to handle - like "Triangle Strip".

Multipatch never really worked in the past either, I'm not sure what to do about it.

File: src/table.jl (hunk 4 above)
Comment (rafaqz) [reply src/table.jl:148]: Well a table is a `FeatureCollection`, so they would go above on `Row`

Comment (rafaqz) [reply src/table.jl:148]: I've added them for `Row`

**Comment by rafaqz** [comment]
Yes apologies about the restructure, it got too tedious scrolling around and finding things with the old structure. A lot of the old code is changed and simplified as well. 

The main new types are `LinearRing`, `LineString` and `SubPolygon` that are lazy objects over the internals of other objects so that GeoInterface.jl methods make sense.

**Comment by rafaqz** [comment]
Merged so we can get ArchGDAL tests passing ASAP. There may be some changes to add to this afterwards, but it should mostly be fine.

**Pull request merged**